File: ml-benchmark/synduce/tree/poly.f
Config NonLinear = true;
Config SampleIntMax = 3;
Config SampleIntMin = -3;
Config SampleSize = 6;

Inductive Tree = nil Unit | node {Int, Tree, Tree};

@Input x: Int;

spec = \t: Tree. (fix (
  \f: {Int, Int} -> Tree -> {Int, Int}. \s: {Int, Int}. \t: Tree.
  match t with
    nil _ -> s
  | node {a, l, r} ->
    let result = (f s l) in
    f {+ result.1 (* result.2 a), * result.2 x} r
  end
) {0,1} t).1;

repr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | node {a, l, r} ->
    node {a, f l, f r}
  end
);

main = \t: Tree. spec (repr t);

File: ml-benchmark/synduce/tree/poly.ml
type tree = Nil of unit | Node of int * tree * tree

let rec spec t =
  let rec f s t =
    match t with
    | Nil _ -> s
    | Node (a, l, r) ->
      let result = f s l in
      match result with
      | (r1, r2) ->
        let new_s = (r1 + r2 * a, r2 * x) in
        f new_s r
  in
  let res = f (0, 1) t in
  match res with
  | (v, _) -> v

let rec repr t =
  match t with
  | Nil _ -> Nil ()
  | Node (a, l, r) -> Node (a, repr l, repr r)

let program t = spec (repr t)